### DISCUSSION TITLE
ZuluSCSI RP2040 support for USB Mass Storage - access SD card contents via USB

### DISCUSSION THREAD

**Discussion by aperezbios** [ZuluSCSI/ZuluSCSI-firmware#346] [Announcements]
Thanks to a firmware contribution from @zigzagjoe, ZuluSCSI RP2040 owners have some experimental/preview firmware that enables USB MASS STORAGE support for ZuluSCSI RP2040, allowing you to access the contents of the SD card via USB. As this is limited to 11Mbit/second USB 1.1 speeds, it is not fast (~700 kilobytes/sec read, ~600ish kB/sec write) but it can be extremely convenient. You can test an experimental build of it out at https://github.com/ZuluSCSI/ZuluSCSI-firmware/releases/tag/v2023.11.29-USBMassStor-BETA

This RP2040-specific firmware works equally with ZuluSCSI Mini (DB25), [ZuluSCSI Compact RP2040](https://store.rabbitholecomputing.com/ZuluSCSI-Compact-RP2040-p/zuluscsi-compact-rp2040.htm) , [ZuluSCSI Compact Homebrew](https://store.rabbitholecomputing.com/ZuluSCSI-Compact-Homebrew-p/zuluscsi-compact-homebrew-kit.htm) kit, as well as the original full-size [ZuluSCSI RP2040 desktop board](https://store.rabbitholecomputing.com/ZuluSCSI-RP2040-p/zuluscsi-rp2040.htm), which Rabbit Hole Computing released a little over a year ago.

**Comment by ZigZagJoe** [comment]
The current UI for card reader mode is simple: if USB is enumerated by a connected computer within ~1 second of power-on, it will enter card reader mode and function as a USB storage. The LED will blink twice quickly then go steady on. It will blink quickly for reads, or slower for writes.

Upon eject from the computer, or if USB isn't enumerated, the led will blink twice then Zulu proceeds to boot up normally as per the current contents of the card. So you can edit zuluscsi.ini and move files around then eject, those settings take effect immediately. The zulu will never enter card reader mode afterwards until you remove power.

Note that booting the attached SCSI device (if any) while in card reader mode will result in no SCSI devices being presented to the attached SCSI device. You will want to eject first. Alternatively, if you do not ever want it to function as a card reader, there's a .INI setting DisableUSBReader to disable it outright. 

(Note: It may make more sense to later make more sense to replace that with a USB reader timeout setting, with 0 being disable.)

**Comment by jforsten** [comment]
If I understood correctly, currently USB access is not possible at the same time as the image is in use? With SCSI2SD v6 it is possible to access the data while card is in use for the SCSI device. This is really usefull for example in Ensoniq samplers where you can easily transfer samples by just writing them to the SD card via USB and then immediately load them in the samper without any extra steps or swap of SD cards.

Is there any possibilty to get that same level of simultaneous access via USB and SCSI to the mounted image with RP2040-based ZuluSCSI?

**Reply by ZigZagJoe** [reply]
Currently, yes, the mass storage mode is mutually exclusive with normal function. So it's either mass storage or a scsi emulator, but not both at the same time. It'd be possible to expose an image file as mass storage rather than the SD card directly, though it'd require a way to configure that and code to make it happen. 

Engineering simultaneous access to an image from both however is much more complicated.... I'm kind of surprised that would have worked, as most filesystems (and USB hosts/SCSI initiators) usually assume that nothing else is touching the drive. Mostly, corruption results if you modify a filesystem that is simultaneously mounted elsewhere. I'm not sure what they'd have done to make that possible.

I personally don't have plans to implement either function, but if you were to do so and submit a pull I'm sure the zulu folks would be receptive.

**Reply by jforsten** [reply]
Thanks for your reply. The USB/SCSI simultaneous access is a default feature with SCSI2SD v6 so as I noticed ZuluSCSI was also using part of v6 source code, I was assuming this woulld be a relatively easy task. Of course the SCSI2SD is working differently as it uses the SD card as a low level drive and there is no extra layer of FAT/exFAT with image file. This would probably make the arbitration easier. I might take a look to the Zulu source to see if there is any reasonable way to implement that. Anyway, thanks for the great project and keep up the good work!

**Reply by aperezbios** [reply]
@jforsten while it may seem like it would be a relatively easy task to someone unfamiliar with the structure of the ZuluSCSI and SCSI2SD V6 code bases, it unfortunately is nowhere near as straightforward as it may seem on the surface, due to internal architectural differences.

 The code that was being leveraged on the SCSI2SD V6 side to enable this functionality was code that is part of [STM32Cube ](https://www.st.com/en/ecosystems/stm32cube.html), which is vendor-supplied, from STMicroelectronics. It's therefore not a component of the SCSI2SD V6 source code which we could use, as it's tightly coupled to the STM32 family of microcontrollers. The portion of SCSI2SD V6 source code that ZuluSCSI utilizes as a library is the portion that handles SCSI command handling, and there are other parts of the SCSI2SD V6 code base which ZuluSCSI does not utilize, for a handful of different reasons. That said, it's certainly a feature we'd like to get implemented, eventually :)

It's one of those "if it were easy/straightforward, it would already be done" sorts of situations. Either way, were we to implement support for what you're requesting, it would probably have to be read-only on the USB side. We'd be happy to consider any pull requests that would enable this sort of functionality, but it isn't as straightforward to implement as it may appear, which is not to say it is impossible.

**Reply by jforsten** [reply]
Thank you for clarification. Maybe from my use case even simpler solution would work. So in order to transfer samples, it would be enough to be able to "eject" the mounted image, access that image from USB side and write the sample file to it, then "attach" the image again and load the sample from sampler side. So what that requires is some way (command via USB?) that would make it possible to change mode from "card reader (USB)" to normal SCSI mode and vice versa. Would that be any more feasible instead of trying to implement simultaneous access to the SD card? 

The reason I'm asking all this is that I would like to extend the support of my tool [ensoniq-remote](https://github.com/jforsten/ensoniq-remote) from SCSI2SD v6 to ZuluSCSI if there would be a way to write sample files to image and then let the Ensoniq to read them (by sending command to Ensoniq via MIDI). I could easily add an extra steps to change the ZuluSCSI mode during that process if there would be a way to do that programmatically.

I have looked also PiSCSI and it sort of supports that idea but it is also more complex system beginning with and has it's own issues.

**Reply by morio** [reply]
@jforsten if the image is contiguous you might have some luck with opening the images as a mass storage device like the SCSI2SD V6. As contiguous images are much like the SCSI2SD architecture where they have a simple start and end sectors defined for the SD card. 
This is the method that will tell you if the image is contiguous: https://github.com/ZuluSCSI/ZuluSCSI-firmware/blob/6cf301e0e7641af4d21bfe0e8b9e542bf390b4e4/src/ImageBackingStore.cpp#L150  

As far as how the SCS2SD v6 code deals with two writers to the same image, you'll have to take a look at the V6 code. I don't have much familiarity with it.

**Comment by kims1000** [comment]
i now have a drive that i cant accces ""drive not found" the sd are in?

no matter what i do, i get the "theres no disk in drive", even if there are sd card in, or if theres no sd card in, i even changed cables??

**Reply by ZigZagJoe** [reply]
Just to be clear, the only time it will enter card reader mode is if you connect it to a PC while the zuluscsi is not otherwise powered (ie. if the device that it is attached to is turned off).

Otherwise it will show up when attached to a computer, but with "no disk in drive" since SCSI is using the SD card.

**Comment by kims1000** [comment]
Mine is  internally in my akai s6000? so i cant turn it off

it works perfekt with the sd card. i can see no ini
i can access the SD card in my akai s6000

what do i do with the no disk in drive

I see the "ZuluSCSI Pico USB Device" in device manger, and it says it works perfectly

I dont see the drive in File Explorer, but only in Total comander 11...why??

**Reply by morio** [reply]
Try the latest release: https://github.com/ZuluSCSI/ZuluSCSI-firmware/releases/tag/v2023.12.14-USBMassStor-BETA

You'll need to edit or create a zuluscsi.ini file on the SD card
Uncomment the line `EnableUSBMassStorage` and set it equal to 1 so `EnableUSBMassStorage = 1`. Under the `[SCSI]` header.

With the sampler turned off, connect the USB cable to the ZuluSCSI, it should mount the SD card as a drive on the computer.

Though it sounds like you are already seeing the device as a drive. It show's up in my Windows 10 File Explorer as `USB Drive (D:)`

At the bottom of this code snippet of the the example zuluscsi.ini you'll see the line you need to uncomment and set to 1.
https://github.com/ZuluSCSI/ZuluSCSI-firmware/blob/f424be987a1cdd8845cf91edf98d292b8092ebb9/zuluscsi.ini#L1-L38

**Comment by ZigZagJoe** [comment]
Turn off the sampler and then plug in the USB to a computer.

**Reply by kims1000** [reply]
sorry still the same: "Drive not found" in Total Commander, and still no drive shown in file explorer - i will try morios suggestions tomorrow morning

ahh yes, i try to be more specific - the drive icon is there in total commander, (not in file explorer) but it does not see the sd card, its like someone have forgotten to put the sd card into the zuluscsi, which i ofcourse havent forgotten :-)

im on win 11

**Comment by kims1000** [comment]
Now i can see the sd card on win 11, and also in file explore, but it freeze my s6000 - its at the so called welcome screen: where it show which version and what is connected (i use v.2.14 btw)

(i can even see the sd card when i take the sd card out of s6000) ???

when i press the buttons and dial on s6000, nothin happens?
i even changed cables?

i restarted s6000 a couple of times, still the same..it works fine again when i dont start with the cable pluckd in

I dont see a v2023.12.14-USBMassStor-BETA, i only see the ZuluSCSI_RP2040_MSC_2023-12-21_f424be9.bin, ok??

this is what i did:
EnableUSBMassStorage = 1 # Enable exposing the SD card as a mass storage device, ok ??
(i removed the #)

**Comment by kims1000** [comment]
...I guess that it is something in the ini, s6000 dont like......dont know

**Reply by morio** [reply]
The way the ZuluSCSI mass storage device works currently, is when it acts as a USB drive, the SCSI side of the board is disabled, so you'll have to keep your sampler turned off. 

You can eject the USB drive from explorer and then turn on your sampler and it should work.
You can also unplug the USB cable and turn your sequencer and the sampler should boot.

I know a lot of people would like to be able to access their files while the host device is on or be able to mount the images as USB devices. But this is at least a start and all the credit getting us this far goes to ZigZagJoe.

**Comment by kims1000** [comment]
Ok thank you for clearing that up :-) so this version pretty much useless for me, i go back to the original version :-)
